### PR TITLE
feat: create staked expenditure

### PR DESCRIPTION
**Pull request by joanna-pagepro**
## Description

Creating a Payment Builder action with the `Staking` decision method, when the Staked expenditure extension is enabled.

## Testing

* Step 1. Make sure that the Staked expenditure extension is installed and enabled and create a Payment builder action with the `Staking` decision method.
* Step 2. When a modal appears, click the `Create payment` button

## Diffs

**New stuff** ✨

* `CreateStakedExpenditureModal` component
* `FinalizeWithStakingInfo` component

**Changes** 🏗

* `ActionButtons` component - `onSubmitClick` prop added to pass the action to show a modal instead of submit
* `ActionSidebarContent`component - added `CreateStakedExpenditureModal` modal
* `DecisionMethodField` and `DecisionMethodRow` components - added `Staking` decision method
* `PaymentBuilderWidget` - if the expenditure is staked `FinalizeWithStakingInfo` is shown instead of `FinalizeWithPermissionsInfo` in `Create` and `Review` steps
* `Button` component - div wrapper added to the loader

**Deletions** ⚰️

* 

## TODO

Contributes to https://github.com/JoinColony/colonyCDapp/issues/2344

### CODE REVIEW

File: src/redux/sagas/expenditures/createStakedExpenditure.ts [hunk 1]
```diff
@@ -239,6 +240,8 @@ function* createStakedExpenditure({
       meta,
     });
 
+    setTxHash?.(txHash);
+
     // @TODO: Remove during advanced payments UI wiring
     // eslint-disable-next-line no-console
     console.log('Created expenditure ID:', expenditureId.toString());
```
Comment (iamsamgibbs) [inline src/redux/sagas/expenditures/createStakedExpenditure.ts:247]: This can be removed now that createStakedExpenditure has been wired up

**Review by iamsamgibbs — APPROVED** [review]
Nice job, this looks good to me!

Screenshot 2024-05-10 at 08 46 51

Screenshot 2024-05-10 at 08 28 32

**Review by marcocolony — CHANGES_REQUESTED** [review]
@joanna-pagepro Could you please have a look at these?

When trying to create a payment with a user that has insufficient locked tokens, the necessary validation is not showing up as per the [designs](https://www.figma.com/file/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?type=design&node-id=23853-196834&mode=design&t=VpKEEaoPkwkovddh-4)

Instead, the modal gets stuck on infinite loading and an error message is displayed in the Userhub.
Export-1715702247339

image

Here's the error being thrown in the console: 
image

File: src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/PaymentBuilderWidget.tsx [hunk 2]
```diff
@@ -91,7 +101,12 @@ const PaymentBuilderWidget: FC<PaymentBuilderWidgetProps> = ({ action }) => {
     {
       key: ExpenditureStep.Create,
       heading: { label: formatText({ id: 'expenditure.createStage.label' }) },
-      content: (
+      content: isStaked ? (
+        <FinalizeWithStakingInfo
```
Comment (jakubcolony) [inline src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/PaymentBuilderWidget.tsx:105]: I believe this component (and related ones) originally displayed motion/expenditure finalization info but it's now used in more places such as creation in this case. Could the naming reflect that better as technically you cannot finalize with staking?

**Review by jakubcolony — CHANGES_REQUESTED** [review]
Great start on this PR @joanna-pagepro.

I noticed the validation error was not shown when I had no tokens activated. There is no horizontal padding on the "Required stake amount" container.
image

The idea behind staking is to allow users with no permissions to create a payment. Currently I'm unable to do that:
image


The create info card itself looks good:
image

My stake was reclaimed after finalizing the payment.

File: src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/PaymentBuilderWidget.tsx (hunk 2 above)
Comment (joanna-pagepro) [reply src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/PaymentBuilderWidget.tsx:102]: Will `StepFinalizedWithStakingInfo` be a better name? If not, maybe you can suggest something?

**Review by marcocolony — CHANGES_REQUESTED** [review]
Hey @joanna-pagepro, thanks for looking at this again. 

I've tried to go through the flow with a member with no permissions. I'm able to create the payment, but when I get to the funding step I'm able to select the permissions decision method and click on fund. Nothing happens after and I'm stuck on this step without an error message explaining why  


Export-1716457088795

image

**Review by jakubcolony — APPROVED** [review]
Code-wise changes look good, nice work @joanna-pagepro 👍

**Comment by arrenv** [comment]
Thank you @joanna-pagepro  and @marcocolony for raising this one.

- Any decision methods used throughout the process should follow the same rules as the action panel decision methods. i.e.
    - Only available ones should show.
    - If no other ones are installed, but the user does not have permissions, then validation should show that they don’t have permissions. With the select disabled. [Figma design link.](https://www.figma.com/design/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?node-id=24483-442557&t=q1re7UWcQs2Bzrf2-4)
    - If reputation is installed, the there is no reputation in the team, then validation should show that there is no reputation to create it. Shown when selecting the Reputation decision method. [Figma design link.](https://www.figma.com/design/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?node-id=24485-442966&t=q1re7UWcQs2Bzrf2-4)

**Comment by joanna-pagepro** [comment]
@marcocolony can you review it again?

**Review by marcocolony — CHANGES_REQUESTED** [review]
Hey @joanna-pagepro thanks for tackling this one again.

I'm still getting the same behaviour, although the console now displays an error stating that the user does not hold the necessary permissions.

image


Export-1716801206190

image

File: src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/PaymentBuilderWidget.tsx (hunk 2 above)
Comment (jakubcolony) [reply src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/PaymentBuilderWidget.tsx:110]: @joanna-pagepro sorry missed your comment, there's no finalization with staking, in this case it's the create step. `CreatedWithStakingInfo` or similar would make sense for this scenario (it's the only action you can do with staking). In other places where `FinalizeWithPermissionsInfo` might be a bit confusing (it's used in cancel, create, review and release steps), maybe something generic such as `ActionWithPermissionsInfo`?

**Comment by joanna-pagepro** [comment]
@marcocolony can you check again? I'm not getting this error. Decision method select is disabled and you can see a box with the info about insufficient permissions.

@jakubcolony naming updated

**Review by marcocolony — CHANGES_REQUESTED** [review]
@joanna-pagepro 

Tested again, the validation is appearing correctly now 🙌 

I am however getting the UI bug where the network badge, user hub and kebab icon are sticking out when the funding modal is active. I think I also reported this on another PR recently? 

Export-1716911330886
image

**Review by arrenv — CHANGES_REQUESTED** [review]
@joanna-pagepro Thank you for putting this one together. Very nice!

1. Reputation was not loading in local dev and it was showing I could create a payment with 0 required tokens. I was able to click the button, but it was stuck in a loading state with the following error in console.

   Can we handle this case by showing an error message in the same red callout card with the text "Reputation in the colony could not loaded, please reload the page and try again." and have the button reload the error on click, instead of trying to trigger the transaction.

image


> I am however getting the UI bug where the network badge, user hub and kebab icon are sticking out when the funding modal is active. I think I also reported this on another PR recently?

@marcocolony this is a feature rather then a bug, when you need to perform a transaction, you need to be able to access these menu items. It is a little bit of a quirk, I think if anything the dark background behind the modal could be removed to avoid the oddness of it.

**Comment by joanna-pagepro** [comment]
@arrenv @marcocolony action updated, can you review again?

**Review by marcocolony — APPROVED** [review]
@joanna-pagepro Nice work, I can see the warning message when reputation isn't loaded in the colony.

image

**Review by arrenv — CHANGES_REQUESTED** [review]
@joanna-pagepro Thank you for the updates.

1. We need the same treatment here for the header to be visible, so that we can interact with the transaction.
image


2. I think there has been some confusion with the activated tokens and validation here, please check this from the spec:
image

      Staking should automatically handle activating tokens, as it does during the motion process, the not enough funds validation is relevant only if they don't have enough funds taking into account their locked tokens, not already staked, and any unlocked funds.

image

image


3. Not sure if introduced in this issue or not, but I don't see it in production, after creating, I can see both Review and Passed status pills in the header. When funding I see passed.
image
image

4. The permission decision method is still visible, despite the users not having permissions, it is not selectable, but it should also not be visible as there is no feedback as to why they cannot select it.
image

**Comment by joanna-pagepro** [comment]
@arrenv 
fixes for 1 and 3 should be already merged to the master or feat/advanced-payments-ui branch
2. This saga throws an error if you don't have enough tokens activated and that's why I implemented it like this. We may need some more changes here. @jakubcolony can you help here?
4. updated

**Comment by joanna-pagepro** [comment]
@arrenv 2 should be also ready to review now

**Comment by joanna-pagepro** [comment]
@arrenv can you check this one? :)

**Review by arrenv — CHANGES_REQUESTED** [review]
@joanna-pagepro Thank you for the update and details.

All else looks good, but I am seeing this issue when funding the payment.

image

**Comment by joanna-pagepro** [comment]
@arrenv should be ok now. Can you check?

**Review by arrenv — APPROVED** [review]
@joanna-pagepro Everything looks good now, thank you for your hard work on this!

**Pull request merged**